velox/exec/RowContainer.cpp

```diff
@@ -385,6 +389,19 @@ int32_t RowContainer::findRows(folly::Range<char**> rows, char** result) {
   return numRows;
 }
 
+void RowContainer::appendNextRow(char* current, char* nextRow) {
+  NextRowVector*& nextRowArrayPtr = getNextRowVector(current);
+  if (!nextRowArrayPtr) {
+    nextRowArrayPtr =
+        new (stringAllocator_->allocate(kNextRowVectorSize)->begin())
```

Comment:
@zhli1142015 is it thread-safe to do so in case of parallel join build? As we might run different partitions from the same table in parallel and string allocator is not thread-safe? cc @Yuhta  Thanks!

Reply:
Since this rarely happen or we don't expect high concurrency here, we could allocate from the memory pool directly which is thread-safe, and we can also simplify the memory reservation mechanism by allocating memory from the last hash build operator which also makes the memory more efficient.

Reply:
Yes it should be. I remember we talked about two possible solution, 1 is to add ThreadSafeStlAllocator. another is to make sure the allocation is from the row cotnainer bound to the thread if in parallel building. Here we use the second way.  Please see: https://github.com/facebookincubator/velox/blob/72c4921d2a7ef00355530ec59dfbf4ec601aff62/velox/exec/HashTable.cpp#L983

Reply:
> Since this rarely happen or we don't expect high concurrency here, we could allocate from the memory pool directly which is thread-safe, and we can also simplify the memory reservation mechanism by allocating memory from the last hash build operator which also makes the memory usage more efficient.

could you please explain more details, I don't get the point here.

Reply:
The parallel join build has two steps:
1. parallel table row partitioning which is one table per thread. It is thread-safe and the partition rows are stored local to each table;
2. parallel table row insertion which is one partition (across all tables) per thread. It is not thread-safe if two threads work on two partitions from the same table. And they allocate memory for duplicate rows from the table's hash string allocator.

I briefly synced with @Yuhta and am working on a solution use memory pool directly to see performance is ok. I guess probe side performance matters here but we don't expect much concurrent memory allocations for the duplicate rows at build side.

Reply:
@zhli1142015 the parallel join insert actually pass the same row container to all the partition insertions which might come from different table. But this might cause the duplicate row vector free code not executed as some subtable's duplicate row flag not set in array mode. Can you help try: https://github.com/facebookincubator/velox/pull/10865 to see if there is regression in your benchmark with real query workloads? Thanks!

Reply:
sure

Reply:
> But this might cause the duplicate row vector free code not executed as some subtable's duplicate row flag not set in array mode.

Thanks @xiaoxmeng  for the fix.

I think besides above fix  there is another change: use memory pool instead of the HashStringAllocator to allocate the memory for the next row vector. From my test, this would cause more memory allocations and slowness. I think this is because we use std::vector here, which needs to free the memory and allocate a bigger buffer(2X always) when scaling up. Because we use HashStringAllocator here, the memory free is not released but stored in the pool of it, then the memory can be reused. So the memory allocation overhead is not so noticeable.
Current
![image](https://github.com/user-attachments/assets/6939aabd-dcdb-487d-a06d-dea716328958)
With your change
![image](https://github.com/user-attachments/assets/99d37218-796e-4c06-bf56-ba727f315f64)

May I know what's the problem we want to solve here?
BTW, One posible solution for above is to use some kind block list to replace vector, like Yuta mentioned in https://github.com/facebookincubator/velox/pull/9079#discussion_r1530664625, if needed I can help on it, thanks.

Reply:
The result is consistent with list join micro-benchmark. I thought there is race condition on HSA but it seems that we already use thread local row container so that's not the problem. But when I test with my change, there is memory leak detected and found out the duplicate row flag in the row container is not always set so the free next row vector is not called on destruction. The current implementation doesn't always allocate the next row vector from its host row container which is a bit confusing. Also to allocate the next row vector from multiple HSAs makes the memory reservation mechanism more complicated and not very efficient (in hash build operator). So I am thinking whether we could directly allocate from the memory pool of the last hash build operator. I am ok with block list if we can allocate from the same memory pool. But since there is no real data race here, it is not urgent and we can revisit later. Thanks!